go/engine/login_username.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2015 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+// LoginUsername is an engine that will get a username or email
+// address from the user and load that user, for the purposes of
+// preparing for provisioning a new device.
+
+package engine
+
+import (
+	"github.com/keybase/client/go/libkb"
+)
+
+// LoginUsername is an engine.
+type LoginUsername struct {
+	libkb.Contextified
+	user            *libkb.User
+	usernameOrEmail string
+}
+
+// NewLoginUsername creates a LoginUsername engine.
+// usernameOrEmail is optional.
+func NewLoginUsername(g *libkb.GlobalContext, usernameOrEmail string) *LoginUsername {
+	return &LoginUsername{
+		Contextified:    libkb.NewContextified(g),
+		usernameOrEmail: usernameOrEmail,
+	}
+}
+
+// Name is the unique engine name.
+func (e *LoginUsername) Name() string {
+	return "LoginUsername"
+}
+
+// GetPrereqs returns the engine prereqs.
+func (e *LoginUsername) Prereqs() Prereqs {
+	return Prereqs{}
+}
+
+// RequiredUIs returns the required UIs.
+func (e *LoginUsername) RequiredUIs() []libkb.UIKind {
+	return []libkb.UIKind{
+		libkb.LoginUIKind,
+		libkb.SecretUIKind,
+	}
+}
+
+// SubConsumers returns the other UI consumers for this engine.
+func (e *LoginUsername) SubConsumers() []libkb.UIConsumer {
+	return nil
+}
+
+// Run starts the engine.
+func (e *LoginUsername) Run(ctx *Context) error {
+	username, err := e.findUsername(ctx)
+	if err != nil {
+		return err
+	}
+
+	e.G().Log.Debug("LoginUsername: found username %q", username)
+
+	user, err := libkb.LoadUser(libkb.NewLoadUserByNameArg(e.G(), username))
+	if err != nil {
+		return e.convertNotFound(err)
+	}
+	e.user = user
+
+	e.G().Log.Debug("LoginUsername: found user %s for username %q", e.user.GetUID(), username)
+
+	return nil
+}
+
+func (e *LoginUsername) User() *libkb.User {
+	return e.user
+}
+
+func (e *LoginUsername) findUsername(ctx *Context) (string, error) {
+	if len(e.usernameOrEmail) == 0 {
+		if err := e.prompt(ctx); err != nil {
+			return "", err
+		}
+	}
+
+	if len(e.usernameOrEmail) == 0 {
+		return "", libkb.NoUsernameError{}
+	}
+
+	if libkb.CheckUsername.F(e.usernameOrEmail) {
+		return e.usernameOrEmail, nil
+	}
+
+	if !libkb.CheckEmail.F(e.usernameOrEmail) {
+		return "", libkb.BadNameError(e.usernameOrEmail)
+	}
+
+	// looks like an email address
+	e.G().Log.Debug("%q looks like an email address, must get login session to get user", e.usernameOrEmail)
+	// need to login with it in order to get the username
+	var username string
+	var afterLogin = func(lctx libkb.LoginContext) error {
+		username = lctx.LocalSession().GetUsername().String()
+		return nil
+	}
+	if err := e.G().LoginState().VerifyEmailAddress(e.usernameOrEmail, ctx.SecretUI, afterLogin); err != nil {
```

Comment:
We can wrap the session transaction around the whole provisioning as well.

I'm not too worried, though...it would be hard to do anything with a bad session file if logged in means that you also have to be provisioned and have a passphrase stream cached.